ml-agents/mlagents/trainers/poca/optimizer_torch.py

```diff
@@ -381,116 +382,110 @@ def _evaluate_by_sequence_team(
-
-        self_seq_obs = []
-        groupmate_seq_obs = []
-        groupmate_seq_act = []
-        seq_obs = []
-        for _self_obs in self_obs:
-            first_seq_obs = _self_obs[0:first_seq_len]
-            seq_obs.append(first_seq_obs)
-        self_seq_obs.append(seq_obs)
-
-        for groupmate_obs, groupmate_action in zip(obs, actions):
-            seq_obs = []
-            for _obs in groupmate_obs:
-                first_seq_obs = _obs[0:first_seq_len]
-                seq_obs.append(first_seq_obs)
-            groupmate_seq_obs.append(seq_obs)
-            _act = groupmate_action.slice(0, first_seq_len)
-            groupmate_seq_act.append(_act)
-
-        # For the first sequence, the initial memory should be the one at the
-        # beginning of this trajectory.
-        for _ in range(first_seq_len):
-            all_next_value_mem.append(ModelUtils.to_numpy(init_value_mem.squeeze()))
-            all_next_baseline_mem.append(
-                ModelUtils.to_numpy(init_baseline_mem.squeeze())
-            )
-
-        all_seq_obs = self_seq_obs + groupmate_seq_obs
-        init_values, _value_mem = self.critic.critic_pass(
-            all_seq_obs, init_value_mem, sequence_length=first_seq_len
-        )
-        all_values = {
-            signal_name: [init_values[signal_name]]
-            for signal_name in init_values.keys()
-        }
-
-        groupmate_obs_and_actions = (groupmate_seq_obs, groupmate_seq_act)
-        init_baseline, _baseline_mem = self.critic.baseline(
-            self_seq_obs[0],
-            groupmate_obs_and_actions,
-            init_baseline_mem,
-            sequence_length=first_seq_len,
-        )
-        all_baseline = {
-            signal_name: [init_baseline[signal_name]]
-            for signal_name in init_baseline.keys()
-        }
+        all_values: Dict[str, List[np.ndarray]] = defaultdict(list)
+        all_baseline: Dict[str, List[np.ndarray]] = defaultdict(list)
+        _baseline_mem = init_baseline_mem
+        _value_mem = init_value_mem
 
         # Evaluate other trajectories, carrying over _mem after each
         # trajectory
         for seq_num in range(
-            1, math.ceil((num_experiences) / (self.policy.sequence_length))
+            0, math.floor((num_experiences) / (self.policy.sequence_length))
         ):
             for _ in range(self.policy.sequence_length):
                 all_next_value_mem.append(ModelUtils.to_numpy(_value_mem.squeeze()))
                 all_next_baseline_mem.append(
                     ModelUtils.to_numpy(_baseline_mem.squeeze())
                 )
 
-            start = seq_num * self.policy.sequence_length - (
-                self.policy.sequence_length - leftover
-            )
-            end = (seq_num + 1) * self.policy.sequence_length - (
-                self.policy.sequence_length - leftover
-            )
+            start = seq_num * self.policy.sequence_length
```

Comment:
This feels like there is a bunch of duplicate code with this part : https://github.com/Unity-Technologies/ml-agents/pull/5206/files#diff-01b42574a4de05de250e29039fc1bbc67b200f3a7c5ed0676b2eee549b943c11R95
Is it possible to combine some of it ?

Reply:
It's annoyingly similar but just slightly different enough to make it really hard to combine. The POCA version has to operate on group obs and actions as well 🤔